lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -402,24 +395,138 @@ bool swift::emitReferenceDependencies(DiagnosticEngine &diags, SourceFile *SF,
+}
+
+static void writeProviders(const SourceFile *SF, llvm::raw_fd_ostream &out) {
+
+  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
+  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
+  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+
+  writeTopLevelProviders(SF->Decls, out, extendedNominals, memberOperatorDecls,
+                         extensionsWithJustMembers);
+
+  writeNominalProviders(extendedNominals, out);
+  writeMemberProviders(extendedNominals, extensionsWithJustMembers, out);
+
+  if (SF->getASTContext().LangOpts.EnableObjCInterop)
+    writeDynamicLookupProviders(SF, out);
+}
+
+static void writeDependencies(const SourceFile *SF,
+                              const DependencyTracker &depTracker,
+                              llvm::raw_fd_ostream &out) {
+  const ReferencedNameTracker *const tracker = SF->getReferencedNameTracker();
+  assert(tracker && "Cannot emit reference dependencies without a tracker");
+
+  auto sortedByName = [](const llvm::DenseMap<DeclBaseName, bool> map)
+      -> SmallVector<std::pair<DeclBaseName, bool>, 16> {
+    SmallVector<std::pair<DeclBaseName, bool>, 16> pairs{map.begin(),
+                                                         map.end()};
+    llvm::array_pod_sort(
+        pairs.begin(), pairs.end(),
+        [](const std::pair<DeclBaseName, bool> *first,
+           const std::pair<DeclBaseName, bool> *second) -> int {
+          return first->first.compare(second->first);
+        });
+    return pairs;
+  };
+
+  writeTopLevelDependencies(sortedByName(tracker->getTopLevelNames()), out);
+
+  {
+    auto &memberLookupTable = tracker->getUsedMembers();
+    std::vector<TableEntryTy> sortedMembers{memberLookupTable.begin(),
+                                            memberLookupTable.end()};
+    llvm::array_pod_sort(
+        sortedMembers.begin(), sortedMembers.end(),
+        [](const TableEntryTy *lhs, const TableEntryTy *rhs) -> int {
+          if (lhs->first.first == rhs->first.first)
+            return lhs->first.second.compare(rhs->first.second);
+
+          if (lhs->first.first->getName() != rhs->first.first->getName())
+            return lhs->first.first->getName().compare(
+                rhs->first.first->getName());
+
+          // Break type name ties by mangled name.
+          auto lhsMangledName = mangleTypeAsContext(lhs->first.first);
+          auto rhsMangledName = mangleTypeAsContext(rhs->first.first);
+          return lhsMangledName.compare(rhsMangledName);
+        });
+
+    writeMemberDependencies(sortedMembers, out);
+    writeNominalDependencies(sortedMembers, out);
   }
 
-  llvm::SmallString<32> interfaceHash;
-  SF->getInterfaceHash(interfaceHash);
-  out << "interface-hash: \"" << interfaceHash << "\"\n";
+  writeExternalDynamicLookupDependencies(
+      sortedByName(tracker->getDynamicLookupNames()), out);
+  writeExternalDependencies(depTracker, out);
+}
+
+namespace swift {
```

Comment:
Why put this in the `swift` namespace?

Reply:
Had to figure out how to call fn with same name, needed "::". Fixed.